# <font color="#704214">PP_005: HIGH VALUE CUSTOMER IDENTIFICATION (INSIDERS)</font>

# <font color='8a795d'>0.0 Settings</font>

## <font color='8a795d'>0.1 Imports</font>

In [8]:
import os
import pickle
import numpy                   as np
import pandas                  as pd
import seaborn                 as sns
import re
import sqlite3
import s3fs
from   sqlalchemy              import create_engine
from   umap                    import umap_         as umap
from   plotly                  import express       as px
from   matplotlib              import pyplot        as plt
from   matplotlib              import cm            as cm
from   sklearn                 import preprocessing as pp
from   sklearn                 import cluster       as cl
from   sklearn                 import metrics       as mt
from   sklearn                 import decomposition as dd
from   sklearn                 import manifold      as mn
from   sklearn                 import ensemble      as en
from   sklearn                 import mixture       as mx
from   sklearn                 import neighbors     as nb
from   scipy.cluster           import hierarchy     as hc
from   scipy                   import stats         as st

In [9]:
# Enviroment variables
AWS_ACCESS_KEY_ID     = os.environ.get( 'AWS_ACCESS_KEY_ID' )
AWS_SECRET_ACCESS_KEY = os.environ.get( 'AWS_SECRET_ACCESS_KEY' )
USER_AWS              = os.environ.get( 'USER_AWS' )
PSWD_AWS              = os.environ.get( 'PSWD_AWS' )
HOST_AWS              = os.environ.get( 'HOST_AWS' )
PORT_AWS              = os.environ.get( 'PORT_AWS' )
DATABASE_AWS          = os.environ.get( 'DATABASE_AWS' )

## <font color='8a795d'>0.3 Loading Dataset</font>

In [11]:
# # load data local
# path_local = '/home/geordano/repos/Insiders_Clustering/insiders_clustering/'
# df_raw = pd.read_csv( path_local + 'data/raw/ecommerce.csv')

# load s3 aws data
path_s3 = 's3://insiders-dataset-geo-ds/'
df_raw = pd.read_csv( path_s3 + 'ecommerce.csv', encoding='iso-8859-1')

# drop extra column
df_raw = df_raw.drop( columns=['Unnamed: 8'], axis=1 )

In [12]:
df_raw.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

  InvoiceDate  UnitPrice  CustomerID         Country  
0   29-Nov-16       2.55     17850.0  United Kingdom  
1   29-Nov-16       3.39     17850.0  United Kingdom  
2   29-Nov-16       2.75     17850.0  United Kingdom  
3   29-Nov-16       3.39     17850.0  United Kingdom  
4   29-Nov-16       3.39     17850.0  United Kingdom

# <font color='8a795d'>1.0 Data Description</font>

In [13]:
#security copy
df1 = df_raw.copy()

## <font color='d2b48c'>1.1 Rename Columns</font>

In [14]:
cols_new = ['invoice_no', 'stock_code', 'description', 'quantity', 
            'invoice_date', 'unit_price', 'customer_id', 'country']
df1.columns = cols_new
df1.head()

invoice_no stock_code                          description  quantity  \
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365      71053                  WHITE METAL LANTERN         6   
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

  invoice_date  unit_price  customer_id         country  
0    29-Nov-16        2.55      17850.0  United Kingdom  
1    29-Nov-16        3.39      17850.0  United Kingdom  
2    29-Nov-16        2.75      17850.0  United Kingdom  
3    29-Nov-16        3.39      17850.0  United Kingdom  
4    29-Nov-16        3.39      17850.0  United Kingdom

## <font color='d2b48c'>1.2 Data Dimensions</font>

In [15]:
print('Number of rows: {}'.format(df1.shape[0]))
print('Number of cols: {}'.format(df1.shape[1]))

Number of rows: 541909
Number of cols: 8


## <font color='d2b48c'>1.3 Data Types</font>

In [16]:
df1.dtypes

invoice_no       object
stock_code       object
description      object
quantity          int64
invoice_date     object
unit_price      float64
customer_id     float64
country          object
dtype: object

## <font color='d2b48c'>1.4 Check NA</font>

In [17]:
df1.isna().sum()

invoice_no           0
stock_code           0
description       1454
quantity             0
invoice_date         0
unit_price           0
customer_id     135080
country              0
dtype: int64

## <font color='d2b48c'>1.5 Replace NA</font>

In [18]:
df_missing = df1.loc[df1['customer_id'].isna(),:]
df_not_missing = df1.loc[~df1['customer_id'].isna(),:]

In [19]:
print('Missing Shape: {}'.format(df_missing.shape))
print('Not Missing Shape: {}'.format(df_not_missing.shape))

Missing Shape: (135080, 8)
Not Missing Shape: (406829, 8)


In [20]:
# Assumption: pode haver invoice_no (compras) replicados nos dois subsets (missing and not missing)
# Explicação: algum erro de sistema pode ter apagado o customer_id de algum invoice_id que se encontra
# no outro subconjunto dos dados.
# Hipótese falsa.

In [21]:
# Assumption: No subset df_missing os invoice_no repetidos se referem a um único customer. 
# Action: replace customer_id = NaN by new id > maximum customer_id of df_not_missing subset
df_not_missing['customer_id'].max()

18287.0

In [22]:
# create reference: invoice_no uniques and sequential customer_id  
df_backup = pd.DataFrame(df_missing['invoice_no'].drop_duplicates())
df_backup['customer_id'] = np.arange(19000, 19000+len(df_backup),1)

# merge original with reference dataframe
df1 = pd.merge(df1, df_backup, on='invoice_no', how='left')

# coalesce
df1['customer_id'] = df1['customer_id_x'].combine_first(df1['customer_id_y'])

# drop extra columns
df1 = df1.drop(columns=['customer_id_x','customer_id_y'], axis=1)

In [23]:
df1.sample(5)

invoice_no stock_code                          description  quantity  \
301593     563348      22728            ALARM CLOCK BAKELIKE PINK         8   
33935      539307      90057       DIAMANTE RING ASSORTED IN BOX.       240   
389678     570488      21677                     HEARTS  STICKERS         2   
78405      542870      22662          LUNCH BAG DOLLY GIRL DESIGN        10   
516406     579899      23562  SET OF 6 RIBBONS PERFECTLY PRETTY           6   

       invoice_date  unit_price         country  customer_id  
301593    13-Aug-17        3.75  United Kingdom      13078.0  
33935     14-Dec-16        1.25  United Kingdom      13694.0  
389678     8-Oct-17        1.63  United Kingdom      14096.0  
78405     30-Jan-17        1.65  United Kingdom      15203.0  
516406    29-Nov-17        2.89  United Kingdom      15687.0

In [24]:
df1.isna().sum()

invoice_no         0
stock_code         0
description     1454
quantity           0
invoice_date       0
unit_price         0
country            0
customer_id        0
dtype: int64

In [25]:
# # remove NA
# df1 = df1.dropna( subset=['description'] )
# print('Removed data: {:.2f}%'.format( ( 1 - ( df1.shape[0] / df_raw.shape[0] ) )*100 ) )

## <font color='d2b48c'>1.6 Change Dtypes</font>

In [26]:
# invoice_date
df1['invoice_date'] = pd.to_datetime( df1['invoice_date'], format='%d-%b-%y' )

# customer_id
df1['customer_id'] = df1['customer_id'].astype( int )

df1.sample(5)

invoice_no stock_code                description  quantity  \
390418     570517      23209   LUNCH BAG VINTAGE DOILY         20   
173083     551652      22083  PAPER CHAIN KIT RETROSPOT        40   
147194     549099      21755   LOVE BUILDING BLOCK WORD         1   
322423     565232      22819  BIRTHDAY CARD, RETRO SPOT         2   
148874     549258      22982        PANTRY PASTRY BRUSH        12   

       invoice_date  unit_price         country  customer_id  
390418   2017-10-09        1.65  United Kingdom        13971  
173083   2017-05-01        2.55          France        12643  
147194   2017-04-04        5.95  United Kingdom        16241  
322423   2017-08-31        0.42  United Kingdom        21619  
148874   2017-04-05        1.25  United Kingdom        13102

In [27]:
df1.dtypes

invoice_no              object
stock_code              object
description             object
quantity                 int64
invoice_date    datetime64[ns]
unit_price             float64
country                 object
customer_id              int64
dtype: object

## <font color='d2b48c'>1.7 Descriptive Statistics</font>

In [28]:
df1.describe().T

count          mean          std       min       25%  \
quantity     541909.0      9.552250   218.081158 -80995.00      1.00   
unit_price   541909.0      4.611114    96.759853 -11062.06      1.25   
customer_id  541909.0  16688.840453  2911.411352  12346.00  14367.00   

                  50%       75%      max  
quantity         3.00     10.00  80995.0  
unit_price       2.08      4.13  38970.0  
customer_id  16249.00  18283.00  22709.0

In [29]:
num_attributes = df1.select_dtypes( include=['int64','float64'])
cat_attributes = df1.select_dtypes( exclude=['int64','float64','datetime64[ns]'])

### <font color='d2b48c'>1.7.1 Numerical Attributes</font>

In [30]:
# central tendency - mean, median
ct1 = pd.DataFrame( num_attributes.apply(np.mean) ).T
ct2 = pd.DataFrame( num_attributes.apply(np.median) ).T

# dispersion - std, min, max, range, skew, kurtosis
d1 = pd.DataFrame( num_attributes.apply( np.min ) ).T
d2 = pd.DataFrame( num_attributes.apply( np.max ) ).T
d3 = pd.DataFrame( num_attributes.apply( lambda x: x.max() - x.min() ) ).T
d4 = pd.DataFrame( num_attributes.apply( np.std ) ).T
d5 = pd.DataFrame( num_attributes.apply( lambda x: x.skew() ) ).T
d6 = pd.DataFrame( num_attributes.apply( lambda x: x.kurtosis() ) ).T

# concatenate
metrics_numerical = pd.concat( [d1, d2, d3, ct1, ct2, d4, d5, d6] ).T.reset_index()
metrics_numerical.columns = ['attributes','min','max','range','mean','median','std','skew','kurtosis']
metrics_numerical

attributes       min      max      range          mean    median  \
0     quantity -80995.00  80995.0  161990.00      9.552250      3.00   
1   unit_price -11062.06  38970.0   50032.06      4.611114      2.08   
2  customer_id  12346.00  22709.0   10363.00  16688.840453  16249.00   

           std        skew       kurtosis  
0   218.080957   -0.264076  119769.160031  
1    96.759764  186.506972   59005.719097  
2  2911.408666    0.487449      -0.804287

### <font color='d2b48c'>1.7.2 Categorical Attributes</font>

In [31]:
cat_attributes.head()

invoice_no stock_code                          description         country
0     536365     85123A   WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
1     536365      71053                  WHITE METAL LANTERN  United Kingdom
2     536365     84406B       CREAM CUPID HEARTS COAT HANGER  United Kingdom
3     536365     84029G  KNITTED UNION FLAG HOT WATER BOTTLE  United Kingdom
4     536365     84029E       RED WOOLLY HOTTIE WHITE HEART.  United Kingdom

#### **Invoice No**

In [32]:
# invoice_no: Há registros com letras e números

# filtrar e contar linhas cujo invoice_no possui letras
df_invoice_letter = df1.loc[df1['invoice_no'].apply( lambda x: bool( re.search('[^0-9]+',x) ) ),: ]
print('invoice_no únicos não numericos: {}'.format(len(df_invoice_letter['invoice_no'].unique())))

# Hipotese: todos os invoice_no correspondem a uma 'quantity' negativa
print('Total de Registros com invoice_no não numericos: {}'.format(len(df_invoice_letter)))
print('Total de Registros com quantity negativa: {}'.format(len(df_invoice_letter[df_invoice_letter['quantity']<0])))

# Assumption: 

invoice_no únicos não numericos: 3839
Total de Registros com invoice_no não numericos: 9291
Total de Registros com quantity negativa: 9288


#### **Stock Code**

In [33]:
# stock_code
# filtrar e contar linhas cujo stock_code possui letras
print('stock_code não numericos: {}'.format(len(df1.loc[df1['stock_code'].apply( 
    lambda x: bool( re.search('[^0-9]+',x) ) ),'stock_code' ])))

# filtrar e contar linhas cujo stock_code possui somente letras
print('stock_code somente letras: {}'.format(len(df1.loc[df1['stock_code'].apply( 
    lambda x: bool( re.search('^[a-zA-Z]+$',x) ) ),'stock_code' ])))

# Quais são os tipos de stock_code que possuem apenas letras
stock_code_list = df1.loc[df1['stock_code'].apply( lambda x: bool( re.search('^[a-zA-Z]+$',x) ) ),'stock_code' ].unique()
print('Tipos de stock_code literais:')
stock_code_list
# Assumption: todos os stock_code apenas literais não representam compras efetivas, portanto serão descartados

stock_code não numericos: 54873
stock_code somente letras: 2759
Tipos de stock_code literais:


array(['POST', 'D', 'DOT', 'M', 'S', 'AMAZONFEE', 'm', 'DCGSSBOY',
       'DCGSSGIRL', 'PADS', 'B', 'CRUK'], dtype=object)

#### **Description**

In [34]:
# Assumption: a feature description é irrelevante para a compreensão do problema de negócio

#### **Country**

In [35]:
# Quantidade de países únicos
print('Quantidade de países únicos: {}'.format(len(df1['country'].unique())))
df1['country'].unique()

# Assumption: European Community e Unspecified não possuem significado como país. Serão excluídos da análise

Quantidade de países únicos: 38


array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [36]:
# Representatividade de cada país no negócio
df_qtd_country = df1['country'].value_counts()
df_perc_country = df1['country'].value_counts(normalize=True)
df_country = pd.concat([df_qtd_country,df_perc_country], axis=1).reset_index()
df_country.columns=['country','qtd_invoice','%_invoice']

# Quantos customers em cada country
df_cust_country = df1[['customer_id','country']].drop_duplicates().groupby('country').count().reset_index()
df_country = pd.merge(df_country, df_cust_country, on='country', how='left').sort_values('customer_id', ascending=False)

In [37]:
df_country.head()

country  qtd_invoice  %_invoice  customer_id
0  United Kingdom       495478   0.914320         7587
1         Germany         9495   0.017521           95
2          France         8557   0.015790           90
3            EIRE         8196   0.015124           44
4           Spain         2533   0.004674           31

# <font color='8a795d'>2.0 Variable Filtering</font>

In [38]:
#security copy
df2 = df1.copy()

In [39]:
# Filtering
# unit_price > 0.04
df2 = df2.loc[df2['unit_price']>=0.04,:]

# invoice_no wrong: same stock_code, same date, same huge ||quantity||
df2 = df2[~df2['invoice_no'].isin(['550461','C550456','570522','C570552','546734','C546735'])]

# stock_code != ['POST','D','DOT','M','S','AMAZONFEE','m','DCGSSBOY','DCGSSGIRL','PADS','B','CRUK']
df2 = df2[~df2['stock_code'].isin(['POST','D','DOT','M','S','AMAZONFEE','m',
                                   'DCGSSBOY','DCGSSGIRL','PADS','B','CRUK'])]

# description
df2 = df2.drop(columns='description', axis=1)

# country - risk
df2 = df2[~df2['stock_code'].isin(['Unspecified','European Community'])]

# bad customers: [16446,12346] huge avg_ticket, but qtd_returned=qtd_purchased
# outlier customer_id:[13672,15823,17548,16546] | item_rp_ratio > 1 (returned >> purchased)
# original customer_id is NaN: [22221,22312,22383] huge quantity of products/items in just one purchase  
df2 = df2[~df2['customer_id'].isin([16446,12346,13672,15823,17548,16546,22221,22312,22383])]

# quantity
df2_returns = df2.loc[df2['quantity']<0,:]
df2_purchases = df2.loc[df2['quantity']>=0,:]

# <font color='8a795d'>3.0 Feature Engineering</font>

In [40]:
#security copy
df3 = df2.copy()
df3_returns = df2_returns.copy()
df3_purchases = df2_purchases.copy()

## <font color='d2b48c'>3.1 Feature Creation</font>

In [41]:
# data reference
df_ref = df3.drop( ['invoice_no', 'stock_code', 'quantity', 'invoice_date', 'unit_price', 'country'], 
                  axis=1 ).drop_duplicates().reset_index( drop=True )

In [42]:
df_ref.shape

(5788, 1)

### <font color='d2b48c'>3.1.1 Gross Revenue</font>

In [43]:
# Gross Revenue
# Purchase Gross Revenue (Faturamento) = quantity * price
df3_purchases.loc[:,'p_gross_revenue'] = df3_purchases.loc[:,'quantity'] * df3_purchases.loc[:,'unit_price']

# Monetary
df_monetary_p = (df3_purchases[['customer_id','p_gross_revenue']]
               .groupby('customer_id')
               .sum()
               .reset_index())


# Return Gross Revenue (Devolução) = quantity * price
df3_returns.loc[:,'r_gross_revenue'] = df3_returns.loc[:,'quantity'] * df3_returns.loc[:,'unit_price']

# Monetary
df_monetary_r = (df3_returns[['customer_id','r_gross_revenue']]
               .groupby('customer_id')
               .sum()
               .reset_index())


# merge
df_ref = pd.merge( df_ref, df_monetary_p, on='customer_id', how='left')
df_ref = pd.merge( df_ref, df_monetary_r, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['p_gross_revenue'].isna(), 'p_gross_revenue'] = 0
df_ref.loc[df_ref['r_gross_revenue'].isna(), 'r_gross_revenue'] = 0
df_ref.isna().sum()

customer_id        0
p_gross_revenue    0
r_gross_revenue    0
dtype: int64

### <font color='d2b48c'>3.1.2 Recency</font>

In [44]:
# Recency (day of the last purchase of the dataset in relation to the day of the last purchase of each customer)
df_recency = (df3_purchases.loc[:,['customer_id','invoice_date']]
              .groupby('customer_id')
              .max()
              .reset_index())
df_recency['recency_days'] = (df3_purchases['invoice_date'].max() - df_recency['invoice_date']).dt.days
df_recency = df_recency[['customer_id','recency_days']].copy()

# Fill NAN with (df3['invoice_date'].max()-df3['invoice_date'].min()).days = 373+1

# merge
df_ref = pd.merge( df_ref, df_recency, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['recency_days'].isna(), 'recency_days'] = 374
df_ref.isna().sum()

customer_id        0
p_gross_revenue    0
r_gross_revenue    0
recency_days       0
dtype: int64

### <font color='d2b48c'>3.1.3 Quantity of Purchases</font>

In [45]:
# Quantity of purchases
df_qtd_purchase = (df3_purchases.loc[:,['customer_id','invoice_no']]
                   .drop_duplicates()
                   .groupby('customer_id')
                   .count()
                   .reset_index()
                   .rename(columns={'invoice_no':'qnt_purchases'}))

# merge
df_ref = pd.merge( df_ref, df_qtd_purchase, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['qnt_purchases'].isna(), 'qnt_purchases'] = 0
df_ref.isna().sum()

customer_id        0
p_gross_revenue    0
r_gross_revenue    0
recency_days       0
qnt_purchases      0
dtype: int64

### <font color='d2b48c'>3.1.4 Variety of Products per Invoice</font>

In [46]:
# Variety of Products per Invoice
df_var_products = (df3_purchases.loc[:,['customer_id','stock_code']]
                   .groupby('customer_id')
                   .agg(tot_stock_code=('stock_code','nunique'))
                   .reset_index()
                   .rename(columns={'tot_stock_code':'var_products'}))

# merge
df_ref = pd.merge( df_ref, df_var_products, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['var_products'].isna(), 'var_products'] = 0
df_ref.isna().sum()

customer_id        0
p_gross_revenue    0
r_gross_revenue    0
recency_days       0
qnt_purchases      0
var_products       0
dtype: int64

### <font color='d2b48c'>3.1.4 Quantity of purchased items per Product</font>

In [47]:
# Quantity of purchased items per product
df_qtd_items = (df3_purchases.loc[:,['customer_id','quantity']]
                .groupby('customer_id')
                .sum()
                .reset_index()
                .rename(columns={'quantity':'qnt_items'}))

# merge
df_ref = pd.merge( df_ref, df_qtd_items, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['qnt_items'].isna(), 'qnt_items'] = 0
df_ref.isna().sum()

customer_id        0
p_gross_revenue    0
r_gross_revenue    0
recency_days       0
qnt_purchases      0
var_products       0
qnt_items          0
dtype: int64

### <font color='d2b48c'>3.1.5 Average Ticket Value</font>

In [48]:
# Average Ticket
df_avg_ticket = (df3_purchases.loc[:,['customer_id','p_gross_revenue']]
                 .groupby('customer_id')
                 .mean()
                 .reset_index()
                 .rename(columns={'p_gross_revenue':'avg_ticket'}))

# merge
df_ref = pd.merge( df_ref, df_avg_ticket, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['avg_ticket'].isna(), 'avg_ticket'] = 0
df_ref.isna().sum()

customer_id        0
p_gross_revenue    0
r_gross_revenue    0
recency_days       0
qnt_purchases      0
var_products       0
qnt_items          0
avg_ticket         0
dtype: int64

### <font color='d2b48c'>3.1.6 Average Recency Days</font> 

In [49]:
# Average Recency Days
df_aux = (df3_purchases.loc[:,['customer_id','invoice_date']]
          .drop_duplicates()
          .sort_values(['customer_id','invoice_date'], ascending=[True,True]))
df_aux['prev_customer_id'] = df_aux['customer_id'].shift()
df_aux['prev_invoice_date'] = df_aux['invoice_date'].shift()

# Recency inter purchases
df_aux['last_recency_days'] = (df_aux.apply( lambda x: ( x['invoice_date'] - x['prev_invoice_date'] )
                                           .days if x['customer_id']==x['prev_customer_id'] else np.nan, axis=1))

# # Fill NaN from customer_id with just one pruchase
# for k in range(len(df_aux)-1):
#     if np.isnan(df_aux.iloc[k,4]):
#         current_id = df_aux.iloc[k,0]
#         next_id = df_aux.iloc[k+1,0]
#         if current_id != next_id:
#             df_aux.iloc[k,4] = (df3['invoice_date'].max() - df_aux.iloc[k,1]).days 

# Average Recency
df_avg_recency_days = (df_aux[['customer_id','last_recency_days']]
                       .groupby('customer_id')
                       .mean()
                       .reset_index()
                       .rename(columns={'last_recency_days':'avg_recency_days'})
                       .dropna())

# merge
df_ref = pd.merge( df_ref, df_avg_recency_days, on='customer_id', how='left')
df_ref.isna().sum()

customer_id            0
p_gross_revenue        0
r_gross_revenue        0
recency_days           0
qnt_purchases          0
var_products           0
qnt_items              0
avg_ticket             0
avg_recency_days    3014
dtype: int64

### <font color='d2b48c'>3.1.7 Frequency Purchase</font>

In [50]:
# Frequency of Purchases
# Tuple Calculus
df_aux = (df3_purchases.loc[:,['customer_id','invoice_no','invoice_date']]
         .drop_duplicates()
         .groupby('customer_id')
         .agg( max_invoice_date=('invoice_date','max'),
               min_invoice_date=('invoice_date','min'),
               days_bought=('invoice_date', lambda x:((x.max()-x.min()).days)+1),
               purchases=('invoice_no','nunique'))
         .reset_index())

# Frequency Purchase
df_aux['freq_purchase'] = df_aux[['purchases','days_bought']].apply( 
    lambda x: x['purchases']/x['days_bought'] if x['days_bought'] != 0 else 0, axis=1 )

df_frequency_purchase = df_aux[['customer_id','freq_purchase']].copy()

# Merge
df_ref = pd.merge( df_ref, df_frequency_purchase, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['freq_purchase'].isna(), 'freq_purchase'] = 0
df_ref.isna().sum()

customer_id            0
p_gross_revenue        0
r_gross_revenue        0
recency_days           0
qnt_purchases          0
var_products           0
qnt_items              0
avg_ticket             0
avg_recency_days    3014
freq_purchase          0
dtype: int64

### <font color='d2b48c'>3.1.8 Number of Returned Items</font>

In [51]:
# Number of Returned Items
df_returns = (df3_returns.loc[:,['customer_id','quantity']]
              .groupby('customer_id')
              .sum()
              .reset_index()
              .rename(columns={'quantity':'qtd_returned'}))
df_returns['qtd_returned'] = df_returns['qtd_returned']*-1

# merge
df_ref = pd.merge( df_ref, df_returns, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['qtd_returned'].isna(), 'qtd_returned'] = 0
df_ref.isna().sum()

customer_id            0
p_gross_revenue        0
r_gross_revenue        0
recency_days           0
qnt_purchases          0
var_products           0
qnt_items              0
avg_ticket             0
avg_recency_days    3014
freq_purchase          0
qtd_returned           0
dtype: int64

### <font color='d2b48c'>3.1.9 Frequency Returns</font>

In [52]:
# Frequency of Returns
# Tuple Calculus
df_aux = (df3_returns.loc[:,['customer_id','invoice_no','invoice_date']]
         .drop_duplicates()
         .groupby('customer_id')
         .agg( max_invoice_date=('invoice_date','max'),
               min_invoice_date=('invoice_date','min'),
               days_return=('invoice_date', lambda x:((x.max()-x.min()).days)+1),
               returns=('invoice_no','nunique'))
         .reset_index())

# Frequency Returns
df_aux['freq_returns'] = df_aux[['returns','days_return']].apply( 
    lambda x: x['returns']/x['days_return'] if x['days_return'] != 0 else 0, axis=1 )

df_returns = df_aux[['customer_id','freq_returns']].copy()

# Merge
df_ref = pd.merge( df_ref, df_returns, on='customer_id', how='left')

# fil NaN's
df_ref.loc[df_ref['freq_returns'].isna(), 'freq_returns'] = 0
df_ref.isna().sum()

customer_id            0
p_gross_revenue        0
r_gross_revenue        0
recency_days           0
qnt_purchases          0
var_products           0
qnt_items              0
avg_ticket             0
avg_recency_days    3014
freq_purchase          0
qtd_returned           0
freq_returns           0
dtype: int64

### <font color='d2b48c'>3.1.10 Average Basket Size</font>

In [53]:
# Average Basket Size
# Tuple Calculus
df_aux = (df3_purchases.loc[:,['customer_id','invoice_no','quantity']]
         .groupby('customer_id')
         .agg( tot_invoice_no=('invoice_no','nunique'),
               tot_quantity=('quantity','sum'))
         .reset_index())

# Average Basket Size: quantity of products on basket
df_aux['avg_basket_size'] = df_aux['tot_quantity']/df_aux['tot_invoice_no']
df_avg_basket = df_aux[['customer_id','avg_basket_size']].copy()

# Merge
df_ref = pd.merge( df_ref, df_avg_basket, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['avg_basket_size'].isna(), 'avg_basket_size'] = 0
df_ref.isna().sum()

customer_id            0
p_gross_revenue        0
r_gross_revenue        0
recency_days           0
qnt_purchases          0
var_products           0
qnt_items              0
avg_ticket             0
avg_recency_days    3014
freq_purchase          0
qtd_returned           0
freq_returns           0
avg_basket_size        0
dtype: int64

### <font color='d2b48c'>3.1.11 Average Basket Variety</font>

In [54]:
# Average Basket Variety
# Tuple Calculus
df_aux = (df3_purchases.loc[:,['customer_id','invoice_no','stock_code']]
         .groupby('customer_id')
         .agg( tot_invoice_no=('invoice_no','nunique'),
               tot_stock_code=('stock_code','nunique'))
         .reset_index())

# Average Basket Variety: variety of products on basket
df_aux['avg_basket_variety'] = df_aux['tot_stock_code']/df_aux['tot_invoice_no']
df_avg_basket_v = df_aux[['customer_id','avg_basket_variety']].copy()

# Merge
df_ref = pd.merge( df_ref, df_avg_basket_v, on='customer_id', how='left')

# fill NaN's
df_ref.loc[df_ref['avg_basket_variety'].isna(), 'avg_basket_variety'] = 0
df_ref.isna().sum()

customer_id              0
p_gross_revenue          0
r_gross_revenue          0
recency_days             0
qnt_purchases            0
var_products             0
qnt_items                0
avg_ticket               0
avg_recency_days      3014
freq_purchase            0
qtd_returned             0
freq_returns             0
avg_basket_size          0
avg_basket_variety       0
dtype: int64

### <font color='d2b48c'>3.1.13 ItemsReturned/ItemsPurchased Ratio</font>

In [55]:
# Quantity of items purchased
df_n_purchases = (df3_purchases.loc[:,['customer_id','quantity']]
              .groupby('customer_id')
              .sum()
              .reset_index()
              .rename(columns={'quantity':'n_items_p'}))


# Number of items Returned
df_n_returns = (df3_returns.loc[:,['customer_id','quantity']]
              .groupby('customer_id')
              .sum()
              .reset_index()
              .rename(columns={'quantity':'n_items_r'}))
df_n_returns['n_items_r'] = df_n_returns['n_items_r']*-1

# Merge two datas
df_aux = pd.merge(df_n_purchases, df_n_returns, on='customer_id', how='left')
df_aux['n_items_r'].fillna(0, inplace=True)

# Returns/Purchases Ratio
df_aux['item_rp_ratio'] = df_aux[['n_items_p','n_items_r']].apply( 
    lambda x: x['n_items_r']/x['n_items_p'] if x['n_items_p'] != 0 else 1, axis=1 )
df_item_rp_ratio = df_aux[['customer_id','item_rp_ratio']].copy()

# Merge
df_ref = pd.merge( df_ref, df_item_rp_ratio, on='customer_id', how='left')

# fill NaN's
df_ref['item_rp_ratio'] = (df_ref[['qtd_returned','item_rp_ratio']].apply( 
                          lambda x: x['qtd_returned'] if np.isnan(x['item_rp_ratio']) 
                          else x['item_rp_ratio'], axis=1 ))
df_ref.isna().sum()

customer_id              0
p_gross_revenue          0
r_gross_revenue          0
recency_days             0
qnt_purchases            0
var_products             0
qnt_items                0
avg_ticket               0
avg_recency_days      3014
freq_purchase            0
qtd_returned             0
freq_returns             0
avg_basket_size          0
avg_basket_variety       0
item_rp_ratio            0
dtype: int64

### <font color='d2b48c'>3.1.14 Percentage Balance (Net Revenue / Gross Revenue)</font>

In [56]:
# percentage billing balance = (p_gross_revenue + r_gross_revenue)/p_gross_revenue
df_ref['perc_balance'] = (df_ref[['p_gross_revenue','r_gross_revenue']].apply( lambda x: 
                          np.round(x['p_gross_revenue'] + x['r_gross_revenue'],3)/x['p_gross_revenue']
                          if x['p_gross_revenue'] != 0 else -1, axis=1 )
                         )
df_ref.isna().sum()

customer_id              0
p_gross_revenue          0
r_gross_revenue          0
recency_days             0
qnt_purchases            0
var_products             0
qnt_items                0
avg_ticket               0
avg_recency_days      3014
freq_purchase            0
qtd_returned             0
freq_returns             0
avg_basket_size          0
avg_basket_variety       0
item_rp_ratio            0
perc_balance             0
dtype: int64

### <font color='d2b48c'>Show</font>

In [57]:
print(df_ref.shape)
df_ref.sort_values('customer_id', ascending=True).head(10)

(5788, 16)


customer_id  p_gross_revenue  r_gross_revenue  recency_days  \
468         12347          4310.00             0.00           2.0   
962         12348          1437.24             0.00          75.0   
5567        12349          1457.55             0.00          18.0   
1691        12350           294.40             0.00         310.0   
1892        12352          1385.74          -120.33          36.0   
3374        12353            89.00             0.00         204.0   
3013        12354          1079.40             0.00         232.0   
3187        12355           459.40             0.00         214.0   
1376        12356          2487.43             0.00          22.0   
5357        12357          6207.67             0.00          33.0   

      qnt_purchases  var_products  qnt_items  avg_ticket  avg_recency_days  \
468             7.0         103.0     2458.0   23.681319         60.833333   
962             4.0          21.0     2332.0   53.231111         94.333333   
5567            1.0          72.0      630.0   20.243750               NaN   
1691            1.0          16.0      196.0   18.400000               NaN   
1892            7.0          57.0      526.0   17.996623         43.333333   
3374            1.0           4.0       20.0   22.250000               NaN   
3013            1.0          58.0      530.0   18.610345               NaN   
3187            1.0          13.0      240.0   35.338462               NaN   
1376            3.0          52.0     1573.0   42.886724        151.500000   
5357            1.0         131.0     2708.0   47.386794               NaN   

      freq_purchase  qtd_returned  freq_returns  avg_basket_size  \
468        0.019126           0.0           0.0       351.142857   
962        0.014085           0.0           0.0       583.000000   
5567       1.000000           0.0           0.0       630.000000   
1691       1.000000           0.0           0.0       196.000000   
1892       0.026820          63.0           1.0        75.142857   
3374       1.000000           0.0           0.0        20.000000   
3013       1.000000           0.0           0.0       530.000000   
3187       1.000000           0.0           0.0       240.000000   
1376       0.009868           0.0           0.0       524.333333   
5357       1.000000           0.0           0.0      2708.000000   

      avg_basket_variety  item_rp_ratio  perc_balance  
468            14.714286       0.000000      1.000000  
962             5.250000       0.000000      1.000000  
5567           72.000000       0.000000      1.000000  
1691           16.000000       0.000000      1.000000  
1892            8.142857       0.119772      0.913166  
3374            4.000000       0.000000      1.000000  
3013           58.000000       0.000000      1.000000  
3187           13.000000       0.000000      1.000000  
1376           17.333333       0.000000      1.000000  
5357          131.000000       0.000000      1.000000

# <font color='8a795d'>4.0 EDA (Exploratory Data Analysis)</font>

In [58]:
# #security copy
df4 = df_ref.dropna().copy()
df4.shape

(2774, 16)

## <font color='d2b48c'>4.3 Space Study</font>

In [78]:
# original dataset
df43 = df4.drop(columns=['customer_id'], axis=1).copy()

In [79]:
# rescale features
# mms = pp.MinMaxScaler()
fs = s3fs.S3FileSystem( anon=False, key=AWS_ACCESS_KEY_ID, secret=AWS_SECRET_ACCESS_KEY)

#p_gross_revenue_scaler    = pickle.load( open( '../features/p_gross_revenue_scaler.pkl', 'rb' ) )#local
p_gross_revenue_scaler     = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/p_gross_revenue_scaler.pkl', 'rb' ) )#aws
df43['p_gross_revenue']    = p_gross_revenue_scaler.transform( df43[['p_gross_revenue']] )

#r_gross_revenue_scaler    = pickle.load( open( '../features/r_gross_revenue_scaler.pkl', 'rb' ) )#local
r_gross_revenue_scaler     = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/r_gross_revenue_scaler.pkl', 'rb' ) )#aws
df43['r_gross_revenue']    = r_gross_revenue_scaler.transform( df43[['r_gross_revenue']] )

#recency_days_scaler       = pickle.load( open( '../features/recency_days_scaler.pkl', 'rb' ) )#local
recency_days_scaler        = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/recency_days_scaler.pkl', 'rb' ) )#aws
df43['recency_days']       = recency_days_scaler.transform( df43[['recency_days']] )

#qnt_purchases_scaler      = pickle.load( open( '../features/qnt_purchases_scaler.pkl', 'rb' ) )#local
qnt_purchases_scaler       = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/qnt_purchases_scaler.pkl', 'rb' ) )#aws
df43['qnt_purchases']      = qnt_purchases_scaler.transform( df43[['qnt_purchases']] )

#var_products_scaler       = pickle.load( open( '../features/var_products_scaler.pkl', 'rb' ) )#local
var_products_scaler        = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/var_products_scaler.pkl', 'rb' ) )#aws
df43['var_products']       = var_products_scaler.transform( df43[['var_products']] )

#qnt_items_scaler          = pickle.load( open( '../features/qnt_items_scaler.pkl', 'rb' ) )#local
qnt_items_scaler           = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/qnt_items_scaler.pkl', 'rb' ) )#aws
df43['qnt_items']          = qnt_items_scaler.transform( df43[['qnt_items']] )

#avg_ticket_scaler         = pickle.load( open( '../features/avg_ticket_scaler.pkl', 'rb' ) )#local
avg_ticket_scaler          = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/avg_ticket_scaler.pkl', 'rb' ) )#aws
df43['avg_ticket']         = avg_ticket_scaler.transform( df43[['avg_ticket']] )

#avg_recency_days_scaler   = pickle.load( open( '../features/avg_recency_days_scaler.pkl', 'rb' ) )#local
avg_recency_days_scaler    = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/avg_recency_days_scaler.pkl', 'rb' ) )#aws
df43['avg_recency_days']   = avg_recency_days_scaler.transform( df43[['avg_recency_days']] )

#freq_purchase_scaler      = pickle.load( open( '../features/freq_purchase_scaler.pkl', 'rb' ) )#local
freq_purchase_scaler       = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/freq_purchase_scaler.pkl', 'rb' ) )#aws
df43['freq_purchase']      = freq_purchase_scaler.transform( df43[['freq_purchase']] )

#qtd_returned_scaler       = pickle.load( open( '../features/qtd_returned_scaler.pkl', 'rb' ) )#local
qtd_returned_scaler        = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/qtd_returned_scaler.pkl', 'rb' ) )#aws
df43['qtd_returned']       = qtd_returned_scaler.transform( df43[['qtd_returned']] )

#freq_returns_scaler       = pickle.load( open( '../features/freq_returns_scaler.pkl', 'rb' ) )#local
freq_returns_scaler        = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/freq_returns_scaler.pkl', 'rb' ) )#aws
df43['freq_returns']       = freq_returns_scaler.transform( df43[['freq_returns']] )

#avg_basket_size_scaler    = pickle.load( open( '../features/avg_basket_size_scaler.pkl', 'rb' ) )#local
avg_basket_size_scaler     = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/avg_basket_size_scaler.pkl', 'rb' ) )#aws
df43['avg_basket_size']    = avg_basket_size_scaler.transform( df43[['avg_basket_size']] )

#avg_basket_variety_scaler = pickle.load( open( '../features/avg_basket_variety_scaler.pkl', 'rb' ) )#local
avg_basket_variety_scaler  = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/avg_basket_variety_scaler.pkl', 'rb' ) )#aws
df43['avg_basket_variety'] = avg_basket_variety_scaler.transform( df43[['avg_basket_variety']] )

#item_rp_ratio_scaler      = pickle.load( open( '../features/item_rp_ratio_scaler.pkl', 'rb' ) )#local
item_rp_ratio_scaler       = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/item_rp_ratio_scaler.pkl', 'rb' ) )#aws
df43['item_rp_ratio']      = item_rp_ratio_scaler.transform( df43[['item_rp_ratio']] )

#perc_balance_scaler       = pickle.load( open( '../features/perc_balance_scaler.pkl', 'rb' ) )#local
perc_balance_scaler        = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/perc_balance_scaler.pkl', 'rb' ) )#aws
df43['perc_balance']       = perc_balance_scaler.transform( df43[['perc_balance']] )

### <font color='d2b48c'>4.3.4 Tree-Based Embedding</font>

In [80]:
# Original Dataset
# y:response variable to be predict | X=training dataset
X = df43.drop(columns=['p_gross_revenue'], axis=1)
y = df43['p_gross_revenue']

# # model definition
# rf_model = en.RandomForestRegressor(n_estimators=100, random_state=42)

# # model training: assemble tree structure
# rf_model.fit(X,y)

# load model
#rf_model = pickle.load( open( '../models/rf_model.pkl', 'rb' ) ) # local
rf_model = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/rf_model.pkl', 'rb' ) )# aws

# dataframe leaf: separation of data in the leaves of the tree
df_leaf = pd.DataFrame(rf_model.apply(X))

In [81]:
#df_tree 2774 customers | n_estimators =100
# Reduce Dimensionality of df_leaf
# reducer = umap.UMAP( n_neighbors=20, random_state=42 )

# load reducer
#reducer = pickle.load( open( '../features/reducer.pkl', 'rb' ) ) # local
reducer = pickle.load( fs.open( 's3://insiders-dataset-geo-ds/reducer.pkl', 'rb' ) )# aws
embedding = reducer.transform( df_leaf )

# embedding
df_tree = pd.DataFrame()
df_tree['embedding_x'] = embedding[:, 0]
df_tree['embedding_y'] = embedding[:, 1]

# <font color='8a795d'>8.0 Model Training</font>

In [83]:
#security copy
X = df_tree.copy()

## <font color='d2b48c'>8.4 DBScan</font>

In [84]:
eps=2.6
min_samples=20

# model definition
dbscan_model = cl.DBSCAN( eps=eps, min_samples=min_samples )

# model training and model predict
labels_dbscan = dbscan_model.fit_predict( X )

# model performance
sil = mt.silhouette_score( X, labels_dbscan, metric='euclidean')

print('Silhouette Score {}'.format( sil ) )
print('Nº of Clusters {}'.format( len( np.unique( labels_dbscan ) ) ) )

Silhouette Score 0.694749653339386
Nº of Clusters 11


# <font color='8a795d'>9.0 Cluster Analysis</font>

## <font color='d2b48c'>9.2 Cluster Profile</font>

In [85]:
# copy of original dataset before rescale
df9_aux = df4.copy()

# set labels from chosen model
df9_aux['cluster'] = labels_dbscan 
df9_aux.head()

customer_id  p_gross_revenue  r_gross_revenue  recency_days  qnt_purchases  \
0        17850          5391.21          -102.58         372.0           34.0   
1        13047          3232.59          -143.49          56.0            9.0   
2        12583          6705.38           -76.04           2.0           15.0   
3        13748           948.25             0.00          95.0            5.0   
4        15100           876.00          -240.90         333.0            3.0   

   var_products  qnt_items  avg_ticket  avg_recency_days  freq_purchase  \
0          21.0     1733.0   18.152222          1.000000      17.000000   
1         105.0     1390.0   18.904035         52.833333       0.028302   
2         114.0     5028.0   28.902500         26.500000       0.040323   
3          24.0      439.0   33.866071         92.666667       0.017921   
4           1.0       80.0  292.000000         20.000000       0.073171   

   qtd_returned  freq_returns  avg_basket_size  avg_basket_variety  \
0          40.0      1.000000        50.970588            0.617647   
1          35.0      0.023973       154.444444           11.666667   
2          50.0      0.105263       335.200000            7.600000   
3           0.0      0.000000        87.800000            4.800000   
4          22.0      0.078947        26.666667            0.333333   

   item_rp_ratio  perc_balance  cluster  
0       0.023081      0.980973        0  
1       0.025180      0.955611        1  
2       0.009944      0.988660        2  
3       0.000000      1.000000        3  
4       0.275000      0.725000        4

In [86]:
# ========== Cluster Profile ========== #
# Number of customers
df_cluster = df9_aux[['customer_id','cluster']].groupby('cluster').count().reset_index()
#df_cluster['perc_customers'] = 100*( df_cluster['customer_id']/df_cluster['customer_id'].sum() )
df_cluster = df_cluster.rename(columns={'customer_id':'customers'})

# Average Gross Revenue
df_avg_gross_rev = df9_aux[['p_gross_revenue','cluster']].groupby('cluster').mean().reset_index()
#df_avg_gross_rev['pot_revenue'] = df_avg_gross_rev['p_gross_revenue']*df_cluster['customers']
#df_avg_gross_rev['perc_revenue'] = 100*( df_avg_gross_rev['pot_revenue']/df_avg_gross_rev['pot_revenue'].sum() )
df_cluster = pd.merge( df_cluster, df_avg_gross_rev, on='cluster', how='inner' )

# Average Recency Days
df_avg_rec_days = df9_aux[['recency_days','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_rec_days, on='cluster', how='inner' )

# Average Qtd Purchases
df_avg_qtd_purchases = df9_aux[['qnt_purchases','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_qtd_purchases, on='cluster', how='inner' )

# Average Qtd Items Purchased
df_avg_qtd_items_purchased = df9_aux[['qnt_items','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_qtd_items_purchased, on='cluster', how='inner' )

# Average Ticket
df_avg_ticket = df9_aux[['avg_ticket','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_ticket, on='cluster', how='inner' )

# Average Frequency Purchase
df_avg_freq_pur = df9_aux[['freq_purchase','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_freq_pur, on='cluster', how='inner' )

# Average Qtd Items Returned
df_avg_qtd_item_ret = df9_aux[['qtd_returned','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_qtd_item_ret, on='cluster', how='inner' )

# Average Frequency Returns
df_avg_freq_ret = df9_aux[['freq_returns','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_freq_ret, on='cluster', how='inner' )

# Average Item Returned/Purchased Ratio
df_avg_rp_ratio = df9_aux[['item_rp_ratio','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_rp_ratio, on='cluster', how='inner' )

# Average Net Margin
df_avg_net_margin = df9_aux[['perc_balance','cluster']].groupby('cluster').mean().reset_index()
df_cluster = pd.merge( df_cluster, df_avg_net_margin, on='cluster', how='inner' )

In [87]:
# Strength: candidate to Insiders 
df_cluster['strength'] = ( (np.log(df_cluster['p_gross_revenue'])
                            *df_cluster['qnt_purchases']
                            *df_cluster['avg_ticket']
                            *df_cluster['perc_balance']) /
                            (df_cluster['recency_days']
                            *df_cluster['freq_returns']
                            *df_cluster['item_rp_ratio']) )

# Índice do Cluster Insiders
insiders = df_cluster.loc[df_cluster['strength'] == df_cluster['strength'].max()]
insiders_cluster = insiders['cluster'].iloc[0]

In [88]:
# DBScan clusters = Best Results
df_cluster.sort_values('p_gross_revenue', ascending=False)#.style.highlight_max(color='lightgreen')

cluster  customers  p_gross_revenue  recency_days  qnt_purchases  \
6         6         36     66079.417222     10.027778      49.722222   
2         2        252      7786.557976     20.746032      15.202381   
5         5        176      3900.387727     27.022727       9.352273   
0         0        157      2902.339045     35.484076       7.611465   
1         1        397      2134.466448     41.730479       5.705290   
7         7        238      1660.996303     32.764706       7.092437   
10       10         68      1375.394706     70.176471       3.102941   
9         9        383      1066.034413     62.872063       2.937337   
3         3        290       838.627069     70.868966       3.286207   
8         8        417       587.750072     78.565947       2.846523   
4         4        360       356.332361     95.391667       2.497222   

       qnt_items  avg_ticket  freq_purchase  qtd_returned  freq_returns  \
6   38902.055556  205.741515       0.141403    962.916667      0.139493   
2    4849.376984   48.566844       0.065514     80.996032      0.217772   
5    2323.420455   38.596523       0.036884     36.562500      0.297291   
0    1791.751592   25.115049       0.139488     28.961783      0.341812   
1    1286.209068   29.976085       0.032461     19.899244      0.329143   
7     778.000000   22.604075       0.031101     17.235294      0.334212   
10    950.235294   21.868676       0.025307      4.514706      0.222824   
9     659.506527   26.353507       0.042479      8.349869      0.280628   
3     429.848276   22.521442       0.039974      5.920690      0.259068   
8     285.779376   20.132083       0.044341      3.597122      0.208602   
4     133.658333   28.947600       0.054236      2.163889      0.167201   

    item_rp_ratio  perc_balance      strength  
6        0.029608      0.966056  2.648368e+06  
2        0.015643      0.980690  9.179677e+04  
5        0.015182      0.977920  2.393091e+04  
0        0.015894      0.982825  7.770716e+03  
1        0.015859      0.980628  5.902270e+03  
7        0.023217      0.974365  4.556109e+03  
10       0.004759      0.993604  6.547112e+03  
9        0.013163      0.981941  2.281702e+03  
3        0.013680      0.980887  1.945800e+03  
8        0.013576      0.981682  1.612228e+03  
4        0.017316      0.976152  1.501256e+03

In [89]:
# Show Cluster Profiles
for k in range(len(df_cluster)):
    if k == insiders_cluster:
        print('\033[1m\033[4m\033[91m'+'Cluster {}'.format(df_cluster['cluster'][k])+' (Insiders)'+'\033[0m')
    else:
        print('\033[1m'+'\033[4m'+'Cluster {}'.format(df_cluster['cluster'][k])+'\033[0m')    
    print('N° de Customers: {}'.format(df_cluster['customers'][k]))#+
           # ' ({:.2f}'.format(df_cluster['perc_customers'][k])+'% dos Customers)')
    print('Receita media: ${:.2f}'.format(df_cluster['p_gross_revenue'][k]))#+
           # ' ({:.1f}'.format(df_cluster['perc_revenue'][k])+'% da Receita)')
    print('Receita Latente: {:.2f}'.format(df_cluster['p_gross_revenue'][k]*df_cluster['customers'][k]))
    print('Recencia media: {:.0f}'.format(df_cluster['recency_days'][k])+' dias')
    print('Frequencia media Compras: {:.2f}'.format(df_cluster['freq_purchase'][k])+' compras/dia')
    print('Qtd media Compras: {:.0f}'.format(df_cluster['qnt_purchases'][k])+' compras')
    print('Qtd media Items Comprados: {:.0f}'.format(df_cluster['qnt_items'][k])+' itens')
    print('Taxa Items Devolvidos: {:.3f}'.format(df_cluster['item_rp_ratio'][k])+' %')
    print('Margem Liquida media: {:.2f}'.format(df_cluster['perc_balance'][k])+' %')
    print('Ticket Medio: ${:.2f}'.format(df_cluster['avg_ticket'][k])+'\n')

Cluster 0
N° de Customers: 157
Receita media: $2902.34
Receita Latente: 455667.23
Recencia media: 35 dias
Frequencia media Compras: 0.14 compras/dia
Qtd media Compras: 8 compras
Qtd media Items Comprados: 1792 itens
Taxa Items Devolvidos: 0.016 %
Margem Liquida media: 0.98 %
Ticket Medio: $25.12

Cluster 1
N° de Customers: 397
Receita media: $2134.47
Receita Latente: 847383.18
Recencia media: 42 dias
Frequencia media Compras: 0.03 compras/dia
Qtd media Compras: 6 compras
Qtd media Items Comprados: 1286 itens
Taxa Items Devolvidos: 0.016 %
Margem Liquida media: 0.98 %
Ticket Medio: $29.98

Cluster 2
N° de Customers: 252
Receita media: $7786.56
Receita Latente: 1962212.61
Recencia media: 21 dias
Frequencia media Compras: 0.07 compras/dia
Qtd media Compras: 15 compras
Qtd media Items Comprados: 4849 itens
Taxa Items Devolvidos: 0.016 %
Margem Liquida media: 0.98 %
Ticket Medio: $48.57

Cluster 3
N° de Customers: 290
Receita media: $838.63
Receita Latente: 243201.85
Recencia media: 71 dias

### Prepare and Export Clusterized Dataset

In [90]:
# Order dataset by 'strength' and save the cluster's index list
cluster_num_list = df_cluster.sort_values('strength', ascending=False)['cluster'].to_list()

# Create alias for all clusters
cluster_alias_list = ['Insiders','A','B','C','D','E','F','G','H','I','J']

# set dictionary with indexes and nicknames from clusters
alias_dict = dict(zip(cluster_num_list, cluster_alias_list))

# set alias column
df9_aux['alias'] = df9_aux['cluster'].map(alias_dict)

# reorder columns
df9_aux = df9_aux[['customer_id','cluster', 'alias', 
                   'p_gross_revenue', 'r_gross_revenue', 'recency_days',
                   'qnt_purchases', 'var_products', 'qnt_items', 
                   'avg_ticket', 'avg_recency_days', 'freq_purchase', 
                   'qtd_returned', 'freq_returns',
                   'avg_basket_size', 'avg_basket_variety', 
                   'item_rp_ratio','perc_balance']]

# # save clusterized dataframe local
# path = '/home/geordano/repos/Insiders_Clustering/insiders_clustering/'
# file_name = path + 'data/processed/ecommerce_deploy.csv'
# df9_aux.to_csv(file_name, encoding='utf-8', index=False)

# # save clusterized dataframe aws
# fs = s3fs.S3FileSystem( anon=False, key=AWS_ACCESS_KEY_ID, secret=AWS_SECRET_ACCESS_KEY)
# file_name_aws = 's3://insiders-dataset-geo-ds/ecommerce_deploy.csv'
# df9_aux.to_csv(file_name_aws, encoding='utf-8', index=False)

# <font color='8a795d'>10.0 EDA for Business</font>

In [91]:
# # load data local
# path = '/home/geordano/repos/Insiders_Clustering/insiders_clustering/'
# df10 = pd.read_csv(path + 'data/processed/ecommerce_deploy.csv')

df10 = df9_aux.copy()

In [92]:
df10.head()

customer_id  cluster alias  p_gross_revenue  r_gross_revenue  recency_days  \
0        17850        0     C          5391.21          -102.58         372.0   
1        13047        1     E          3232.59          -143.49          56.0   
2        12583        2     A          6705.38           -76.04           2.0   
3        13748        3     H           948.25             0.00          95.0   
4        15100        4     J           876.00          -240.90         333.0   

   qnt_purchases  var_products  qnt_items  avg_ticket  avg_recency_days  \
0           34.0          21.0     1733.0   18.152222          1.000000   
1            9.0         105.0     1390.0   18.904035         52.833333   
2           15.0         114.0     5028.0   28.902500         26.500000   
3            5.0          24.0      439.0   33.866071         92.666667   
4            3.0           1.0       80.0  292.000000         20.000000   

   freq_purchase  qtd_returned  freq_returns  avg_basket_size  \
0      17.000000          40.0      1.000000        50.970588   
1       0.028302          35.0      0.023973       154.444444   
2       0.040323          50.0      0.105263       335.200000   
3       0.017921           0.0      0.000000        87.800000   
4       0.073171          22.0      0.078947        26.666667   

   avg_basket_variety  item_rp_ratio  perc_balance  
0            0.617647       0.023081      0.980973  
1           11.666667       0.025180      0.955611  
2            7.600000       0.009944      0.988660  
3            4.800000       0.000000      1.000000  
4            0.333333       0.275000      0.725000

## <font color='d2b48c'>10.2 Business Hypothesis</font>

#### H1: Os clientes do grupo *Insiders* participam com mais de 10% do total de vendas
+ **VERDADEIRA:** O grupo *Insiders* participa com 10.66% do volume total de vendas 

In [72]:
# DBSCAN
# soma da quantidade de compras do grupo Insiders
df_purchases_insiders = df10.loc[df10['alias']=='Insiders', 'qnt_purchases'].sum()

# quantidade total de compras
df_purchases_total = df10.loc[:,'qnt_purchases'].sum()

# proporção das compras do grupo Insiders em relação ao total de compras
print(f'Percentual das compras do Insiders: {100*df_purchases_insiders/df_purchases_total:.2f} % do Total')

Percentual das compras do Insiders: 10.66 % do Total


#### H2: Os clientes do grupo *Insiders* contribuem com mais de 25% do faturamento total
+ **VERDADEIRA:** O grupo *Insiders* contribui com 30.33% do faturamento total de vendas

In [73]:
# DBSCAN
# soma do faturamento bruto de compras do grupo Insiders
df_revenue_insiders = df10.loc[df10['alias']=='Insiders', 'p_gross_revenue'].sum()

# faturamento bruto total de compras
df_revenue_total = df10.loc[:,'p_gross_revenue'].sum()

# proporção do faturamento do grupo Insiders em relação ao total de compras
print(f'Percentual do faturamento do Insiders: {100*df_revenue_insiders/df_revenue_total:.2f} % do Total')

Percentual do faturamento do Insiders: 30.33 % do Total


#### H3: Os clientes do grupo *Insiders* possuem *Basket Size Médio* maior do que a média da base total de clientes
+ **VERDADEIRA:** O grupo *Insiders* possui Basket Size Medio (1124.21) maior que a média da base (229.41)

In [74]:
# DBSCAN
# Basket Size Medio do grupo Insiders
avg_basket_size_insiders = df10.loc[df10['alias']=='Insiders', 'avg_basket_size'].mean()

# Basket Size Medio da base total de clientes
avg_basket_size_total = df10['avg_basket_size'].mean()

# teste da condição enunciada na hipótese
if( avg_basket_size_insiders > avg_basket_size_total ):
    print('True:  O Basket Size Medio do Insiders ({:.2f}) é maior do que o Basket Size Medio da base ({:.2f})'
          .format(avg_basket_size_insiders,avg_basket_size_total))
else:
    print('False:  O Basket Size Medio do Insiders ({:.2f}) é menor do que o Basket Size Medio da base ({:.2f})'
          .format(avg_basket_size_insiders,avg_basket_size_total))

True:  O Basket Size Medio do Insiders (1124.21) é maior do que o Basket Size Medio da base (229.41)


#### H4: O grupo *Insiders* possui, em média, *Average Ticket* maior do que a média do restante da base total de clientes
+ **VERDADEIRA:** O grupo *Insiders* possui *Average Ticket* Medio (205.74) maior que a média do restante da base (28.19)

In [75]:
# DBSCAN
# Media do Average Ticket do grupo Insiders
mean_avg_ticket_insiders = df10.loc[df10['alias']=='Insiders', 'avg_ticket'].mean()

# Media do Average Ticket do grupo Insiders
mean_avg_ticket_remaining = df10.loc[df10['alias']!='Insiders', 'avg_ticket'].mean()

# teste da condição enunciada na hipótese
if( mean_avg_ticket_insiders > mean_avg_ticket_remaining ):
    print('True:  A média do Average Ticket do Insiders ({:.2f}) \n é maior do que a média do Average Ticket do restante da base ({:.2f})'
          .format(mean_avg_ticket_insiders,mean_avg_ticket_remaining))
else:
    print('False:  A média do Average Ticket do Insiders ({:.2f}) \n é menor do que a média do Average Ticket do restante da base ({:.2f})'
          .format(mean_avg_ticket_insiders,mean_avg_ticket_remaining))

True:  A média do Average Ticket do Insiders (205.74) 
 é maior do que a média do Average Ticket do restante da base (28.19)


#### H5: O percentil 70 do Faturamento Bruto do grupo *Insiders* é maior que a média do grupo
+ **FALSA:** O faturamento médio do cluster *Insiders* é maior do que seu percentil 70  
(70% dos clientes do cluster Insiders possuem Faturamento Bruto menor do que a média do grupo)

In [76]:
# COMPARE
perc_70_g_rev_insiders = np.percentile(df10.loc[df10['alias']=='Insiders'].p_gross_revenue, q=70)
mean_g_rev_insiders = df10.loc[df10['alias']=='Insiders'].p_gross_revenue.mean()

print('Percentil 70 do Faturamento do Insiders: {:.2f}'.format(perc_70_g_rev_insiders))
print('Média do Faturamento do Insiders: {:.2f}'.format(mean_g_rev_insiders))

Percentil 70 do Faturamento do Insiders: 59796.87
Média do Faturamento do Insiders: 66079.42


## <font color='d2b48c'>10.3 Business Questions</font>

### <font color='ggg777'>*1. Quem são as pessoas elegíveis para participar do programa de Insiders ?*</font>

In [77]:
# DBSCAN 36
insiders_customer_list = df10.loc[df10['alias']=='Insiders', 'customer_id'].to_list()
print(insiders_customer_list)

[15311, 16029, 17511, 13408, 13694, 12748, 14911, 17841, 13777, 17381, 15061, 14156, 13798, 14680, 16013, 17949, 15769, 13081, 13089, 16422, 17450, 15838, 18102, 17857, 14298, 17404, 16684, 12931, 14646, 13027, 12415, 14088, 13098, 16333, 12901, 14096]


### <font color='ggg777'>*2. Quantos clientes farão parte do grupo?*</font>

In [78]:
# DBSCAN 36
print(f'Total de Clientes que farão parte do grupo Insiders: {len(insiders_customer_list)} ({100*len(insiders_customer_list)/df10.customer_id.size:.2f} % da base)')

Total de Clientes que farão parte do grupo Insiders: 36 (1.30 % da base)


### <font color='ggg777'>*3. Quais as principais características desses clientes ?*</font>

In [79]:
# DBSCAN 36 customers
# get statistics from cluster Insiders by describe method
df_insiders_statistics = df10.loc[df10['alias']=='Insiders'].describe().drop(columns=['customer_id','cluster'])

# get mean of all features
df_insiders_profile = df_insiders_statistics.T.iloc[:,1].apply(lambda x: np.round(x,2))

# Get features'names
features_names_list = df_insiders_statistics.columns.to_list()

# Create alias for all features
features_alias_list = ['Faturamento(compras) Médio',
                       'Faturamento(retornos) Médio',
                       'Recência Média',
                       'Quantidade Média de Compras',
                       'Variedade Média de Produtos Comprados',
                       'Quantidade Média de Items Comprados',
                       'Média do Ticket Médio',
                       'Média da Recência entre compras',
                       'Frequência Média de Compras',
                       'Quantidade Média de Itens Devolvidos',
                       'Frequência Média de Devoluções',
                       'Média do Basket Size Médio',
                       'Média do Basket Variety Médio',
                       'Percentual Médio de Itens Devolvidos',
                       'Percentual Médio de Saldo Faturado']

# set dictionary with indexes and nicknames from features
alias_feat_dict = dict(zip(features_names_list, features_alias_list))

# rename columns of series with alias dictionary
df_insiders_profile.rename(index=alias_feat_dict)

Faturamento(compras) Médio               66079.42
Faturamento(retornos) Médio              -1914.10
Recência Média                              10.03
Quantidade Média de Compras                 49.72
Variedade Média de Produtos Comprados      366.64
Quantidade Média de Items Comprados      38902.06
Média do Ticket Médio                      205.74
Média da Recência entre compras             16.14
Frequência Média de Compras                  0.14
Quantidade Média de Itens Devolvidos       962.92
Frequência Média de Devoluções               0.14
Média do Basket Size Médio                1124.21
Média do Basket Variety Médio                8.85
Percentual Médio de Itens Devolvidos         0.03
Percentual Médio de Saldo Faturado           0.97
Name: mean, dtype: float64

### <font color='ggg777'>*4. Qual a porcentagem de contribuição do faturamento vinda do Insiders ?*</font>

In [80]:
# DBSCAN 36 customers
# soma da quantidade de compras do grupo Insiders
df_purchases_insiders = df10.loc[df10['alias']=='Insiders', 'p_gross_revenue'].sum()

# quantidade total de compras
df_purchases_total = df10.loc[:,'p_gross_revenue'].sum()

# proporção das compras do grupo Insiders em relação ao total de compras
print(f'Percentual do faturamento do Insiders: {100*df_purchases_insiders/df_purchases_total:.2f} % do Total')

Percentual do faturamento do Insiders: 30.33 % do Total


### <font color='ggg777'>*5. Qual a expectativa de faturamento desse grupo para os próximos meses ?*</font>

 This question cannot be answered because clustering did not take time into account. 
 
 The method was applied throughout the available time frame, disregarding the behavior over time.

### <font color='ggg777'>*6. Quais as condições para uma pessoa ser elegível ao Insiders ?*</font>

#### DBSCAN 36 customers

In [81]:
# Faixa de faturamento de confiança para participar do Insiders
df_insiders_revenue = df10.loc[df10['alias']=='Insiders', 'p_gross_revenue']

# intervalo de confiança de variabilidade em torno da média
min_rev, max_rev = st.t.interval(alpha=0.95, 
                                 df=len(df_insiders_revenue), 
                                 loc=np.mean(df_insiders_revenue), 
                                 scale=st.sem(df_insiders_revenue))
print(f'Faixa de faturamento para ser elegível ao Insiders: ${np.round(min_rev,2)} a ${np.round(max_rev,2)}')

Faixa de faturamento para ser elegível ao Insiders: $44836.16 a $87322.68


In [82]:
# Faixa de recencia de confiança para participar do Insiders
df_insiders_recency = df10.loc[df10['alias']=='Insiders', 'recency_days']

# intervalo de confiança de variabilidade em torno da média
min_rev, max_rev = st.t.interval(alpha=0.95, 
                                 df=len(df_insiders_recency), 
                                 loc=np.mean(df_insiders_recency), 
                                 scale=st.sem(df_insiders_recency))
print(f'Faixa de recência para ser elegível ao Insiders: {int(np.round(min_rev,0))} a {int(np.round(max_rev,0))} dias')

Faixa de recência para ser elegível ao Insiders: 3 a 17 dias


In [83]:
# Faixa de frequencia de confiança para participar do Insiders
df_insiders_frequency = df10.loc[df10['alias']=='Insiders', 'freq_purchase']

# intervalo de confiança de variabilidade em torno da média
min_rev, max_rev = st.t.interval(alpha=0.95, 
                                 df=len(df_insiders_frequency), 
                                 loc=np.mean(df_insiders_frequency), 
                                 scale=st.sem(df_insiders_frequency))
print(f'Faixa de frequência para ser elegível ao Insiders: {np.round(min_rev,3)} a {np.round(max_rev,3)} compras/dia')

Faixa de frequência para ser elegível ao Insiders: 0.101 a 0.182 compras/dia


### <font color='ggg777'>*7. Quais as condições para uma pessoa ser removida do Insiders ?*</font>

Apresentar performance abaixo da faixa de elegibilidade das métricas do Insiders

# <font color='8a795d'>11.0 Deploy To Production</font>

In [93]:
# Security Copy and change dtypes
df11 = df9_aux.copy()

# Change types
df11['recency_days'] = df11['recency_days'].astype(int)
df11['qnt_purchases'] = df11['qnt_purchases'].astype(int)
df11['var_products'] = df11['var_products'].astype(int)
df11['qnt_items'] = df11['qnt_items'].astype(int)
df11['avg_recency_days'] = df11['avg_recency_days'].astype(int)
df11['qtd_returned'] = df11['qtd_returned'].astype(int)

## <font color='d2b48c'>11.1 Insert into Database</font>

In [94]:
# database connection local

# path_local = '/home/geordano/repos/Insiders_Clustering/insiders_clustering/src/data/'
# endpoint_local = 'sqlite:///'+path_local+'insiders_db.sqlite'

# # CREATE TABLE

# query_create_table_insiders = """
#     CREATE TABLE insiders (
#         customer_id         INTEGER,
#         cluster             INTEGER,
#         alias               TEXT,
#         p_gross_revenue     REAL,
#         r_gross_revenue     REAL,
#         recency_days        INTEGER,
#         qnt_purchases       INTEGER,
#         var_products        INTEGER,
#         qnt_items           INTEGER,
#         avg_ticket          REAL,
#         avg_recency_days    INTEGER,
#         freq_purchase       REAL,
#         qtd_returned        INTEGER,
#         freq_returns        REAL,
#         avg_basket_size     REAL,
#         avg_basket_variety  REAL,
#         item_rp_ratio       REAL,
#         perc_balance        REAL
#     )
# """

# # create connection local
# conn = sqlite3.connect(path_local+'insiders_db.sqlite')
# conn.execute(query_create_table_insiders)
# conn.commit()
# conn.close()

# conn = create_engine('sqlite:///'+path_local+'insiders_db.sqlite') # local
# df11.to_sql('insiders', con=conn, if_exists='append', index=False)

### <font color='d2b48c'>11.1.1 Insert into MariaDB Database</font>

In [95]:
# #----Using MariaDB---

# connection parameters
user     = USER_AWS
pswd     = PSWD_AWS
host     = HOST_AWS
port     = PORT_AWS
database = DATABASE_AWS
dbtype   = 'mysql+pymysql' # MariaDB

endpoint = dbtype+'://'+user+':'+pswd+'@'+host+':'+port+'/'+database

# create engine 
engine     = create_engine( endpoint )
print('Engine>>>> {}'.format(engine))

# create connection
connection = engine.connect()
print('Connection>>>> {}'.format(connection))

Engine>>>> Engine(mysql+pymysql://insiders_aws_geo:***@database-insiders.ct75wow3b7wj.us-east-2.rds.amazonaws.com:3306/database_insiders)
Connection>>>> <sqlalchemy.engine.base.Connection object at 0x7f9c8c100520>


In [96]:
# DROP TABLE

# drop table query
query_drop_insiders = """
    DROP TABLE insiders
"""

# execute drop query
connection.execute(query_drop_insiders);

In [97]:
# CREATE TABLE

query_create_table_insiders = """
    CREATE TABLE insiders (
        customer_id         INTEGER,
        cluster             INTEGER,
        alias               TEXT,
        p_gross_revenue     REAL,
        r_gross_revenue     REAL,
        recency_days        INTEGER,
        qnt_purchases       INTEGER,
        var_products        INTEGER,
        qnt_items           INTEGER,
        avg_ticket          REAL,
        avg_recency_days    INTEGER,
        freq_purchase       REAL,
        qtd_returned        INTEGER,
        freq_returns        REAL,
        avg_basket_size     REAL,
        avg_basket_variety  REAL,
        item_rp_ratio       REAL,
        perc_balance        REAL
    )
"""

connection.execute(query_create_table_insiders)

In [98]:
# Insert Data

#----Using Engine----
try:
    df11.to_sql('insiders', con=connection, if_exists='append', index=False)
except Exception as error1:
    print("Error with connection ------> ")
    print(error1)

In [99]:
# # Select Data
# query = """
#     SELECT * FROM insiders
# """

# df_select = pd.read_sql_query( query, connection)
# df_select

customer_id  cluster alias  p_gross_revenue  r_gross_revenue  \
0           17850        0     C          5391.21          -102.58   
1           13047        1     E          3232.59          -143.49   
2           12583        2     A          6705.38           -76.04   
3           13748        3     H           948.25             0.00   
4           15100        4     J           876.00          -240.90   
...           ...      ...   ...              ...              ...   
2769        13596        3     H           697.04             0.00   
2770        14893        9     G          1237.85             0.00   
2771        14126        9     G           706.13           -62.50   
2772        13521        9     G          1092.39             0.00   
2773        15060        8     I           301.84             0.00   

      recency_days  qnt_purchases  var_products  qnt_items  avg_ticket  \
0              372             34            21       1733   18.152222   
1               56              9           105       1390   18.904035   
2                2             15           114       5028   28.902500   
3               95              5            24        439   33.866071   
4              333              3             1         80  292.000000   
...            ...            ...           ...        ...         ...   
2769             5              2           133        406    4.199036   
2770             9              2            72        799   16.956849   
2771             7              3            14        508   47.075333   
2772             1              3           312        733    2.511241   
2773             8              4            80        262    2.515333   

      avg_recency_days  freq_purchase  qtd_returned  freq_returns  \
0                    1      17.000000            40      1.000000   
1                   52       0.028302            35      0.023973   
2                   26       0.040323            50      0.105263   
3                   92       0.017921             0      0.000000   
4                   20       0.073171            22      0.078947   
...                ...            ...           ...           ...   
2769                 7       0.250000             0      0.000000   
2770                 2       0.666667             0      0.000000   
2771                 3       0.750000            50      1.000000   
2772                 4       0.300000             0      0.000000   
2773                 1       2.000000             0      0.000000   

      avg_basket_size  avg_basket_variety  item_rp_ratio  perc_balance  
0           50.970588            0.617647       0.023081      0.980973  
1          154.444444           11.666667       0.025180      0.955611  
2          335.200000            7.600000       0.009944      0.988660  
3           87.800000            4.800000       0.000000      1.000000  
4           26.666667            0.333333       0.275000      0.725000  
...               ...                 ...            ...           ...  
2769       203.000000           66.500000       0.000000      1.000000  
2770       399.500000           36.000000       0.000000      1.000000  
2771       169.333333            4.666667       0.098425      0.911489  
2772       244.333333          104.000000       0.000000      1.000000  
2773        65.500000           20.000000       0.000000      1.000000  

[2774 rows x 18 columns]

In [100]:
# close connection
connection.close()

### <font color='d2b48c'>11.1.2 Insert into PostgreSQL Database</font>

In [74]:
# # database connection aws

# import psycopg2

# # parameters PostgreSQL
# user     = 'user'
# pswd      = 'pswd!'
# host     = 'host'
# port     = '5432'
# database = 'postgres' 
# dbtype   = 'postgresql' # PostgreSQL

# endpoint = dbtype+'://'+user+':'+pswd+'@'+host+':'+port+'/'+database
# #print('Endpoint = {}'.format(endpoint))

# # create engine 
# engine     = create_engine( endpoint )
# print('Engine>>>> {}'.format(engine))

# # create connection
# connection = engine.connect()
# print('Connection>>>> {}'.format(connection))

# # DROP TABLE
# # database connection
# # conn = create_engine('sqlite:///'+path+'insiders_db.sqlite')

# # drop table
# query_drop_insiders = """
#     DROP TABLE insiders_ggg
# """

# # execute drop query
# connection.execute(query_drop_insiders);

# # CREATE TABLE

# query_create_table_insiders = """
#     CREATE TABLE insiders_ggg (
#         customer_id         INTEGER,
#         cluster             INTEGER,
#         alias               TEXT,
#         p_gross_revenue     REAL,
#         r_gross_revenue     REAL,
#         recency_days        INTEGER,
#         qnt_purchases       INTEGER,
#         var_products        INTEGER,
#         qnt_items           INTEGER,
#         avg_ticket          REAL,
#         avg_recency_days    INTEGER,
#         freq_purchase       REAL,
#         qtd_returned        INTEGER,
#         freq_returns        REAL,
#         avg_basket_size     REAL,
#         avg_basket_variety  REAL,
#         item_rp_ratio       REAL,
#         perc_balance        REAL
#     )
# """

# connection.execute(query_create_table_insiders)

# # Insert Data
# #----Using Engine----
# try:
#     df11.to_sql('insiders_ggg', con=conn, if_exists='append', index=False)
# except Exception as error1:
#     print("Error with connection ------> ")
#     print(error1)

# # Select Data
# query = """
#     SELECT * FROM insiders_ggg
# """

# df_select = pd.read_sql_query( query, conn)
# df_select

# connection.close()